VideoPlayer updates

FernetMenta opened the pull request:
This adds video buffers to be used by decoders and renderers. platforms can register custom buffers like dma.
render formats are history, renderers need to check type of video buffer

platforms register 
- decoders 
- renderers
- processInfo

FernetMenta committed:
VideoPlayer: rename and move YuvImage

FernetMenta committed:
VideoPlayer: drop CodecUtils::Allocat/FreePicture

FernetMenta committed:
VideoPlayer: cleanup CodecUtils

FernetMenta committed:
VideoPlayer: add video buffers

lrusak committed:
[cmake] allow building gbm with vaapi

FernetMenta committed:
follow ffmpeg docs for avio_alloc_context

FernetMenta committed:
VideoPlayer: add some logging for audio

FernetMenta committed:
VideoPlayer: move video queue level to ProcessInfo

popcornmix committed:
omxplayer: Move GetLevel to ProcessInfo

popcornmix committed:
MMAL: Use CVideoBuffer and IVideoBufferPool as bases for MMAL

FernetMenta committed:
VideoPlayer: create a state for gui/video render

FernetMenta committed:
VideoPlayer: drop render formats

FernetMenta committed:
VideoPlayer: OSX adapt to video buffers

FernetMenta committed:
VideoPlayer: change videoPicture to const

FernetMenta committed:
VideoPlayer: drop PreInit, move supported pix formats to PprocessInfo

FernetMenta committed:
VideoPlayer: vdpau - add runtime check

FernetMenta committed:
RP compile error

FernetMenta committed:
VideoPlayer: vaapi - adapt to video buffers

FernetMenta committed:
VideoPlayer: vdpau - adapt to VB

popcornmix committed:
aml / linuxgles adaption

popcornmix committed:
RenderManager: Protect against null videoBuffer

popcornmix committed:
RenderManager: MMALRenderer should always be used when available

popcornmix committed:
PixelConverterRBP: Fixup

popcornmix committed:
LinuxRendererGLES: Make it compile

popcornmix committed:
OMXPlayerVideo: fixups

popcornmix committed:
MMAL: Fixups

garbear commented on xbmc/cores/VideoPlayer/Process/ProcessInfo.h:
```diff
@@ -110,4 +114,5 @@ class CProcessInfo
   // player states
   CCriticalSection m_stateSection;
   bool m_stateSeeking;
+  std::atomic_int m_levelVQ;
```
The default constructor of `std::atomic_int` leaves the object in an uninitialized state. Calling GetLevelVQ() before SetLevelVQ() will result in UB. `m_levelVQ` should be initialized by `CProcessInfo` or the UB should be clearly documented.

FernetMenta replied:
> Calling GetLevelVQ() before SetLevelVQ()

that can't happen, even without docu

garbear replied:
What about if RP doesn't use levels and doesn't call SetLevelVQ()? does the GUI know not to call GetLevelVQ()?

FernetMenta replied:
RP must not use VP's private APIs. I told you many times already. VP takes care that VQ represents actual state of MQ

garbear replied:
is CProcessInfo a private interface? I assumed otherwise because PAPlayer uses it

FernetMenta replied:
everything not in the top level folder of VideoPlayer is not supposed to be used by other components. CProcessInfo represents the state and behavior of VP that is totally different to the one of RP

garbear replied:
Understood. One of the outcomes of the GSoC project will be to remove all uses of VP code from retroplayer.

FernetMenta replied:
that would be best for now but does not mean that we can't factor out common code at a later time. as long VP's code base has not stabilised I think that is best strategy.

garbear commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/MMALRenderer.h:
```diff
@@ -94,8 +94,14 @@ class CMMALRenderer : public CBaseRenderer, public CThread, public IRunnable
+  // Player functions
+  virtual bool         Configure(unsigned int width, unsigned int height, unsigned int d_width, unsigned int d_height, float fps, unsigned flags, ERenderFormat format, unsigned extended_format, unsigned int orientation);
+  virtual int          GetImage(YuvImage *image, int source = AUTOSOURCE, bool readonly = false);
+>>>>>>> 50fb06e... VideoPlayer: rename and move YuvImage
```
Oops, rebase error

FernetMenta replied:
intentional, has to be touched anyway

garbear commented on xbmc/cores/VideoPlayer/VideoPlayer.h:
```diff
@@ -402,14 +400,16 @@ class CVideoPlayer : public IPlayer, public CThread, public IVideoPlayer, public
   virtual void UpdateRenderInfo(CRenderInfo &info) override;
   virtual void UpdateRenderBuffers(int queued, int discard, int free) override;
+  virtual void UpdareGuiRender(bool gui) override;
+  virtual void UpdareVideoRender(bool video) override;
```
Small typo

FernetMenta replied:
yes, already corrected in one of my branches and will update tomorrow

garbear commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTBGL.cpp:
```diff
@@ -56,38 +51,31 @@ CRenderInfo CRendererVTB::GetRenderInfo()
+        vb->m_fence = 0;
+      }
+    }
+    vb->Release();
```
this should be moved 1 line above, right?

FernetMenta replied:
that entire block needs rework. testing that dynamic cast makes no sense in ReleaseBuffer. if it were no vtb type, this would not have been added.

garbear committed:
VideoPlayer: Fix missing #include

garbear committed:
VideoPlayer: Fix missing virtual destructors

garbear committed:
RetroPlayer: Update for video buffers

garbear commented:
I submitted some fixes in #12220 and updated RP in #12221.

FernetMenta committed:
VideoPlayer: fix wrong values in DataCacheCore

garbear committed:
RetroPlayer: Fix RPi after videobuffers update

garbear committed:
RetroPlayer: Fix spelling error

popcornmix committed:
MMAL: Remove unused decoder parameter

garbear committed:
VideoPlayer: Remove unused function

garbear committed:
VideoPlayer: Move trivial implementations to superclass

popcornmix committed:
MMAL: Move to MMAL namespace

Also tidy includes and move CMMALBuffer to renderer

peak3d commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1645,17 +1611,54 @@ void CVaapiRenderPicture::GLUnMapSurface()
+class VAAPI::CVaapiBufferPool : public IVideoBufferPool
+{
+public:
+  CVaapiBufferPool(CDecoder &decoder);
```
What is CDecoder here? MemoryBuffer is MemoryBuffer and should not rely on any thing like Decoder.
From theory anything could write into a buffer, not only decoders,

peak3d commented:
From my understanding BufferPools must be part of the platform, not decoder (codec).
RetroPlayer / AddonVideoCodec must not know anything about VAAPI or DMA or whatever.

They should request an buffer from any global place and write to it without knowing whats the underlying buffer-type. Why shouldn't RetoPlayer be able to write into an mmapped VABuffer?

Hopefully I missed something when looking over the code, from (short!!) reading I can only say that the concept does not match the needs I have with AddonVideoCodec (and what I thought is optimal for RetroPlayer).

FernetMenta commented:
every component can own the bufferpool. it depends on the requiremetns. in case of vaapi, vaapi owns the pool. you are free to register any kind of pool with the bufferPoolManager for use in AddonVideoCodec.

> Why shouldn't RetoPlayer be able to write into an mmapped VABuffer?

you cannot map vaapi surface to system mem. hence RP won't ever be able to write into a vaapi surface.

Rechi commented on xbmc/cores/DataCacheCore.h:
```diff
@@ -67,6 +67,9 @@ class CDataCacheCore
+  void SetGuiRender(bool gui);
+  bool GetGuiRender();
+  void SetVideoRender(bool video);
 
```
`bool GetVideoRender();` is missing

FernetMenta replied:
somehow got lost on rebase, fixed now

popcornmix committed:
DVDVideoCodecFFmpeg: Clear videoBuffer after release

Rechi commented on xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp:
```diff
@@ -0,0 +1,386 @@
+  uint8_t *d = pDst->plane[0];
+  int w = pDst->width * pDst->bpp;
+  int h = pDst->height;
+  if ((w == pSrc->stride[0]) && ((unsigned int) pSrc->stride[0] == pDst->stride[0]))
```
casting not needed

Rechi replied:
why casting to `unsigned int` here?
`pSrc->stride[1]` and `pDst->stride[1]` have the same type
same is done multiple times below

FernetMenta replied:
I just copied the old code: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/DVDCodecUtils.cpp#L118

right, cast is not needed anymore

FernetMenta replied:
updated, @Rechi  thanks for bringing this up

FernetMenta committed:
VideoPlayer: drop some obsolete casts, squash to videobuffers

Rechi commented on xbmc/cores/VideoPlayer/Process/VideoBuffer.cpp:
```diff
@@ -0,0 +1,386 @@
+  int h = pDst->height;
+
+  // Copy YUYV
+  if ((w * 2 == pSrc->stride[0]) && ((unsigned int) pSrc->stride[0] == pDst->stride[0]))
```
@FernetMenta you missed one superfluous cast

pkerling committed:
Move VAAPI display creation to WinSystem

popcornmix committed:
VideoBuffers: Use YuvImage::MAX_PLANES

afedchin committed:
[videobuffer] adapt windows

afedchin commented:
windows changes are done (not tested)

FernetMenta commented:
great, thanks!

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAHD.h:
```diff
@@ -104,6 +105,7 @@ class CProcessorHD : public ID3DResource
   D3D11_VIDEO_PROCESSOR_RATE_CONVERSION_CAPS m_rateCaps;
   D3D11_TEXTURE2D_DESC m_texDesc;
   PROCESSOR_VIEW_TYPE m_eViewType;
+  std::vector<AVPixelFormat> m_formats;
```
this is obsolete, formates have to be supplied by CProcessInfo::GetRenderFormats

afedchin replied:
how I can decide in CProcessInfo::GetRenderFormats wich format is supported by dxva renderer if I don't know will it be used or not?

afedchin replied:
there is CWinRenderer::GetRenderInfo() where I provide supported formats.

FernetMenta replied:
>  if I don't know will it be used or not

this problem exists in current master.

in this branch a decoder delivers a videobuffer and renderer decides if it can handle it. capabilites don't change at runtime, means this is very static behaviour.  in case it depends on runtime environemnt, you can ask windowing in CProcessInfo::GetRenderFormats

afedchin replied:
what I should ask windowing? is setting enabled && can dxvahd be inited && will HW decoder used?

FernetMenta replied:
m_formats only ever had a single purpose: to tell sw decoder if it needs to convert the output of ffmpeg decoder. a supported format should not depend on any settings. 
nevertheless you can query settings in windows' implementation of CProcessInfo::GetRenderFormats. but if you need that, something goes really wrong.

afedchin replied:
In case of sw decoding and DXVA render method enabled with yuv420p10 video. DXVA may not support yuv420p10 format, actually it is possible to detect supported fornat after initing DXVA renderer
it's not a problem to drop it, but I think it will be a pain

FernetMenta replied:
this is static information you can collect at system start. it won't change during runtime, hence no need to overcomplicate startup of playback. take this as an example: https://github.com/FernetMenta/kodi-agile/blob/videobuffer/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp#L1195
this is called at startup by windowing. you could do the same with your renderer

afedchin replied:
> it won't change during runtime

different adapters has different caps. I think restart app after changing adapter it more worst idea than do not check caps at all.

FernetMenta replied:
simply call the static method again after adapters have changed. I hooked vaapi/vdpau into RefreshContext: https://github.com/FernetMenta/kodi-agile/blob/videobuffer/xbmc/windowing/X11/WinSystemX11GLContext.cpp#L213

afedchin commented:
now windows code works as expected (I think)

afedchin committed:
[dxvahd] cosmetic changes

FernetMenta commented:
thanks much. I'll rebase and squash your changes end of this week when I'll have vdpau changes ready.

popcornmix committed:
VideoBuffer: Avoid unsafe race when resetting pool

Currently we are resetting the m_pool variable after Returning the buffer.
As the pool will add the buffer back to the free list there is a danger it
will be reallocated and m_pool assigned before the reset to nullptr occurs.

That leaves you with a live buffer with an invalid pool pointer.
So, make sure we don't touch m_pool after returning the buffer.

popcornmix committed:
VideoBuffer: Fix unsafe use of atomics

The separate accesses of m_refCount breaks atomicity.
If m_refCount==2 and two threads both call realease, they can both decrement first
and then both call pool->Return which is unwanted.

Access m_refCount atomically to avoid this.

Rechi commented:
@FernetMenta this needs a rebase

FernetMenta committed:
VideoPlayer: have platforms register their laform specific codecs into factory

FernetMenta committed:
VideoPlayer: make AddonVideoCodec compile - fix me

FernetMenta committed:
VideoPlayer: vaapi - factor out egl interop

FernetMenta committed:
VideoPlayer: vdpau - add register codec

FernetMenta committed:
X11: register vdpau and vaapi

FernetMenta commented:
@peak3d AddonVideoCodec needs update

popcornmix committed:
MMAL: Register HW codecs from windowing

FernetMenta committed:
VideoPlayer: osx - register vtb codec

popcornmix committed:
MMAL: Pool should be created with unknown encoding type

popcornmix committed:
MMALRender: Specified aligned width/height should be forced

popcornmix committed:
AddonVideoCodec: Switch to using CVideoBufferManager

peak3d committed:
AMLVideoBufferPool

FernetMenta committed:
VideoPlayer: introduce factory for renderers

FernetMenta committed:
VideoPlayer: add renderFormats for LinuxRendererGL

FernetMenta committed:
VideoPlayer: vaapi - move egl interop check to renderer

FernetMenta committed:
X11: register renderers

FernetMenta committed:
OSX: register renderers

FernetMenta committed:
windows: register decoder and renderer

FernetMenta commented:
@popcornmix I introduced a factory for renderers. renderers need to be registered like decoders. see x11, windows, osx for examples

Rechi commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VTB.h:
```diff
@@ -19,24 +19,44 @@
 #pragma once
 
-#include "system_gl.h"
+#include <OpenGL/gl.h>
```
iOS need `OpenGLES/ES2/gl.h` instead of `OpenGL/gl.h`

FernetMenta replied:
updated

peak3d committed:
Render registration

peak3d committed:
Remove Bypass / Render videoBuffer in LinuxRendererGLES

peak3d committed:
Remove unused pbo / set VideoBuffer::loaded

Rechi commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VTB.h:
```diff
@@ -19,24 +19,48 @@
+#include <OpenGL/gl.h>
+#else
+#include <OpenGLES/ES2/gl.h>
+#endif
```
this change isn't enough, have a look at https://github.com/xbmc/xbmc/blob/f743b10/xbmc/system.h#L189-L193
system.h undefines `HAS_GL` for iOS

peak3d committed:
Fix LinuxRendererGLES / Register AMLCodec

peak3d committed:
Fix refcounting

popcornmix committed:
MMAL: Register hardware renderer

FernetMenta committed:
VideoPlayer: register ProcessInfo

FernetMenta commented:
@afedchin @popcornmix @peak3d the concept of having platforms register their components is now also applied to ProcessInfo 5a2d60e

peak3d committed:
Android MediaCodec:: VideoBuffer implementation

peak3d committed:
AudioDecoderHW Registration / DVDCodecAudioAndroidMediaCodec

popcornmix committed:
MMAL: Move mmal buffer initialisation to update function

peak3d committed:
Fix FactoryCodec::Audio / MediaCodecOES

lrusak committed:
gbm: register components

peak3d committed:
Register LinuxRenderGLES / Fix MediaCodec Non-Surface

FernetMenta committed:
VideoPlayer: ios - adapt to changes

FernetMenta commented:
@Memphiz ios added. I did a quick test with vtb and sw decoding

Memphiz commented:
Thx for doing all the Darwin work!

popcornmix committed:
RBP: Register ProcessInfo

popcornmix committed:
CVideoBufferManager: Add critical section

FernetMenta committed:
RP: fix compile

FernetMenta committed:
VideoPlayer: make VideoBufferManager non copyable

FernetMenta committed:
VideoPlayer: vdpau - register renderer

FernetMenta committed:
VideoPlayer: vtbgl - register renderer

FernetMenta committed:
VideoPlayer: LinuxRendererGL - clear backbuffer if render failed

FernetMenta committed:
VideoPlayer: interface cleanups, override

FernetMenta committed:
windows: cosmetics, whitepaces

FernetMenta committed:
windows: register ProcessInfo

FernetMenta commented:
jenkins build this please

MartijnKaijser commented:
@popcornmix rpi fails

popcornmix commented:
Yep, I'll PR a fix

FernetMenta commented:
jenkins build this please

FernetMenta commented:
jenkins build this please

MartijnKaijser commented:
fyi playback seems fine on android

FernetMenta commented:
@MartijnKaijser thanks. Do we have volunteers for runtime tests on Windows?

afedchin commented:
I've tested very quickly while adapting windows changes.

FernetMenta commented:
@afedchin could you please review 1b5e4966ebb66215be2228e172026835ee42adc3

FernetMenta committed:
VideoPlayer: rework VideoBuffers

hudokkow committed:
[modernize][Videoplayer] Prefer 'override' usage

FernetMenta committed:
VideoPlayer: make dealing with vb refs safer

Rechi commented on xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp:
```diff
@@ -19,29 +19,46 @@
+    if (ret)
+      return ret;
+  }
+  if (!ret)
```
why doing this check here?
This produces a warning that the function end may be reached without a `CProcessInfo*` return.

FernetMenta replied:
right, I'll drop it

peak3d committed:
Pass currentClock in Renderer::AddVideoPicture

FernetMenta committed:
VideoPlayer: windows - vb fixes

peak3d committed:
[Android] ReleaseBufferAtTime() for MediaCodecSurface

FernetMenta commented:
jenkins build this please

FernetMenta commented:
jenkins build this please

MartijnKaijser commented:
Last I build rpi didn't finish. http://jenkins.kodi.tv/job/LINUX-RBPI/4883/

FernetMenta commented:
how can we make jenkins build this again?

MartijnKaijser commented:
After removing the label it needs a while to adapt. 
jenkins build this please

FernetMenta commented:
jenkins build this please

lrusak committed:
[cmake] add FindOpenGLES3.cmake

FernetMenta commented:
as far as I am concered this is ready to go now. is anybody aware of remaining issues that should be addressed prior to merge?

garbear commented on xbmc/cores/VideoPlayer/DVDDemuxers/DemuxMultiSource.h:
```diff
@@ -45,22 +45,21 @@ class CDemuxMultiSource : public CDVDDemux
   void Flush() override;
-  std::string GetFileName() override { return ""; };
   int GetNrOfStreams() const override;
-  CDemuxStream* GetStream(int iStreamId)const override { return nullptr; } ;
```
Uh oh, this doesn't look good. In the superclass this is pure-virtual. I think I moved this to `protected` access but it got lost in the rebase. I'll try running a test build and see if I'm correct.

peak3d committed:
[AML][Android] compile / runtime fixes

lrusak committed:
OpenGLES: update for OpenGLES3

lrusak committed:
[cmake] gbm requires opengles3 when vaapi is enabled

FernetMenta commented:
jenkins build this please

FernetMenta commented:
jenkins build this please

lrusak committed:
VideoShaders: replace confusing ifdefs

FernetMenta committed:
VideoPlayer: vaapi - fix hevc interop check

lrusak committed:
VideoPlayer Prefer default for declarations

MartijnKaijser commented:
let's get it in

FernetMenta commented:
@lrusak wanted to submit gbm updates in the morning (his timezone)

lrusak commented:
gbm works fine with vaapi for me :+1:

FernetMenta commented:
jenkins build this please

FernetMenta commented:
win64 error unrelated.

dapharsyde commented:
Sorry I know I'm late to the party, but I just discovered that this merge seems to be the cause of stuttering 1080p live TV streams.  720p is fine.  I have confirmed it is not network related, this issue is only present when I upgrade my build from version #0709 to #0710.  I am using VDPAU acceleration, in case that is relevant.   Can anyone help?

popcornmix commented:
I imagine creating a [trac](https://trac.kodi.tv/) ticket would have a better chance of a response than posting on an old github PR.
